### PR TITLE
EOS-19731: Consul separation: replace hare-consul-agent.service with consul.service

### PR DESCRIPTION
**Pull request by mssawant**
Presently Hare installs a separate Consul systemd unit, hare-consul-agent which
starts a Consul agent and configures the same. Instead the default Consul systemd
unit can be used. This will help decouple Hare from Consul and use it as a 3rd party software.

Solution:
- Provide a `skip-consul` option to `hctl bootstrap` command, this will help Hare
  to support both the options, use default Consul systemd unit or hare-consul-agent
  systemd unit so that the deployments not using Consul as a separate service need
  not have to start and configure Consul separately.
- Remove dependency of `hare-consul-agent` systemd unit from `hare-hax` systemd unit.
- As default Consul service data dir can be different the functions used by the
  watchers need to use the absolute paths, e.g. `get-service-health` used in `check-service` watcher script.

Note:
- Default Consul service uses `consul` as a user and group, this must be configured
  and added to sudoers list for watcher scripts to work.
- Default Consul service configuration is not part of these code changes.

### CODE REVIEW

**Comment by mssawant** [comment]
Bootstrap without `--skip-consul`,
```
[root@ssc-vm-1623:root] hctl bootstrap --mkfs /root/cluster-3-1-ios.yaml
2021-06-21 22:40:27: Generating cluster configuration... OK
2021-06-21 22:40:30: Starting Consul server on this node... OK
2021-06-21 22:40:30: Waiting for Consul's internal leader to be ready... OK
2021-06-21 22:40:30: Importing configuration into the KV store... OK
2021-06-21 22:40:31: Starting Consul on other nodes... OK
2021-06-21 22:40:31: Updating Consul configuraton from the KV store... OK
2021-06-21 22:40:37: Waiting for the RC Leader to get elected... OK
2021-06-21 22:40:38: Starting Motr (phase1, mkfs)... OK
2021-06-21 22:40:57: Starting Motr (phase1, m0d)... OK
2021-06-21 22:40:59: Starting Motr (phase2, mkfs)... OK
2021-06-21 22:41:14: Starting Motr (phase2, m0d)... OK
2021-06-21 22:41:19: Checking health of services... OK
```

Bootstrap with `--skip-consul`,
```
[root@ssc-vm-1623:root] hctl bootstrap --skip-consul --mkfs /root/cluster-3-1-ios.yaml
2021-06-21 22:43:10: Generating cluster configuration... OK
2021-06-21 22:43:13: Checking Consul server on this node... OK
2021-06-21 22:43:13: Waiting for Consul's internal leader to be ready... OK
2021-06-21 22:43:13: Importing configuration into the KV store... OK
2021-06-21 22:43:14: Checking Consul on other nodes... OK
2021-06-21 22:43:15: Updating Consul configuraton from the KV store... OK
2021-06-21 22:43:16: Waiting for the RC Leader to get elected... OK
2021-06-21 22:43:16: Starting Motr (phase1, mkfs)... OK
2021-06-21 22:43:30: Starting Motr (phase1, m0d)... OK
2021-06-21 22:43:38: Starting Motr (phase2, mkfs)... OK
2021-06-21 22:43:56: Starting Motr (phase2, m0d)... OK
2021-06-21 22:44:01: Checking health of services... OK
[root@ssc-vm-1623:root] hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x7b 'pool_0'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xc5 'default': 'pool_0' None None
Services:
    ssc-vm-1624.colo.seagate.com
    [started]  hax        0x7200000000000001:0x2e  10.230.250.53@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x31  10.230.250.53@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0x34  10.230.250.53@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x4d  10.230.250.53@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x50  10.230.250.53@tcp:12345:4:2
    ssc-vm-1623.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x6   10.230.250.51@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x9   10.230.250.51@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0xc   10.230.250.51@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x25  10.230.250.51@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x28  10.230.250.51@tcp:12345:4:2
    ssc-vm-1625.colo.seagate.com
    [started]  hax        0x7200000000000001:0x56  10.230.250.48@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x59  10.230.250.48@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0x5c  10.230.250.48@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x75  10.230.250.48@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x78  10.230.250.48@tcp:12345:4:2
```

**Comment by andriytk** [comment]
`skip-consul` is a bad and misleading name: we still use Consul, we don't skip it, but we use the system one, right? Suggestion: name it `use-system-consul` or something.

Also:
>Remove dependency of hare-consul-agent systemd unit from hare-hax systemd unit.

Why?

Also: if we are going to use default system Consul - why do we need hare-consul? In which scenarios?

**Comment by mssawant** [comment]
@andriytk,
>Also: if we are going to use default system Consul - why do we need hare-consul? In which scenarios? I'd rather suggest choosing and use only one approach: either to use default system Consul or hare-consul. And don't support both - it's redundant and hard to maintain.

Yes, that will be indeed simpler. Thought it would help having an option where may be some users (SAGE, community) don't need to learn and put effort into configuring (basic) Consul separately. But we can always document a how to.

**Comment by andriytk** [comment]
But Hare must configure Consul also, right? So why user should configure the default system one?

**Comment by mssawant** [comment]
@andriytk, by configuring I mean setting the required parameters in the default `/etc/consul.d/consul.hcl`, Hare will not do this. It will be user's responsibility to start consul cluster before starting Hare. Hare will just be another user of the Consul cluster and bootstrap will fail if Consul cluster is not running already.
That is the reason I preserved the existing version of Hare that will start and configure Consul cluster using `hare-consul-agent` iff user need not start a separate Consul cluster if its not used outside of Hare.

**Comment by mssawant** [comment]
Mini provisioner,
```
[root@ssc-vm-1623:root] /opt/seagate/cortx/hare/bin/hare_setup config --config "json:///root/hare.config.conf.3-node" --file '/root/cluster-3-1-ios-consul.yaml'
[root@ssc-vm-1623:root]

[root@ssc-vm-1623:root] /opt/seagate/cortx/hare/bin/hare_setup init --config "json:///root/hare.config.conf.3-node" --file '/root/cluster-3-1-ios-consul.yaml'
Cluster is not running
2021-06-23 14:04:17: Generating cluster configuration... OK
2021-06-23 14:04:21: Checking Consul server on this node... OK
2021-06-23 14:04:21: Waiting for Consul's internal leader to be ready... OK
2021-06-23 14:04:22: Importing configuration into the KV store... OK
2021-06-23 14:04:48: Checking Consul on other nodes... OK
2021-06-23 14:04:48: Updating Consul configuraton from the KV store... OK
2021-06-23 14:04:50: Waiting for the RC Leader to get elected... OK
2021-06-23 14:04:50: Starting Motr (phase1, mkfs)... OK
2021-06-23 14:05:44: Starting Motr (phase1, m0d)... OK
2021-06-23 14:05:55: Starting Motr (phase2, mkfs)... OK
2021-06-23 14:07:51: Starting Motr (phase2, m0d)... OK
2021-06-23 14:08:38: Starting S3 servers (phase3)... OK
2021-06-23 14:08:39: Checking health of services... OK
2021-06-23 14:08:42,109 [INFO] Waiting for all the processes to start..
2021-06-23 14:08:44,114 [INFO] Waiting for all the processes to start..
2021-06-23 14:08:46,121 [INFO] Waiting for all the processes to start..
2021-06-23 14:08:48,127 [INFO] Waiting for all the processes to start..
2021-06-23 14:08:50,133 [INFO] Waiting for all the processes to start..
2021-06-23 14:08:52,140 [INFO] Waiting for all the processes to start..
2021-06-23 14:08:54,803 [INFO] Waiting for all the processes to start..
2021-06-23 14:08:56,809 [INFO] Waiting for all the processes to start..
2021-06-23 14:08:58,814 [INFO] Waiting for all the processes to start..
2021-06-23 14:09:00,821 [INFO] Waiting for all the processes to start..
2021-06-23 14:09:02,827 [INFO] Waiting for all the processes to start..
Stopping s3server@0x7200000000000001:0x2e at ssc-vm-1623.colo.seagate.com...
Stopping s3server@0x7200000000000001:0x31 at ssc-vm-1623.colo.seagate.com...
Stopping s3server@0x7200000000000001:0x68 at ssc-vm-1624.colo.seagate.com...
Stopping s3server@0x7200000000000001:0x6b at ssc-vm-1624.colo.seagate.com...
Stopping s3server@0x7200000000000001:0xa2 at ssc-vm-1625.colo.seagate.com...
Stopping s3server@0x7200000000000001:0xa5 at ssc-vm-1625.colo.seagate.com...
Stopped s3server@0x7200000000000001:0x2e at ssc-vm-1623.colo.seagate.com
Stopped s3server@0x7200000000000001:0xa2 at ssc-vm-1625.colo.seagate.com
Stopped s3server@0x7200000000000001:0xa5 at ssc-vm-1625.colo.seagate.com
Stopped s3server@0x7200000000000001:0x68 at ssc-vm-1624.colo.seagate.com
Stopped s3server@0x7200000000000001:0x6b at ssc-vm-1624.colo.seagate.com
Stopped s3server@0x7200000000000001:0x31 at ssc-vm-1623.colo.seagate.com
Stopping m0d@0x7200000000000001:0x1a (ios) at ssc-vm-1623.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x9 (ios) at ssc-vm-1623.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x43 (ios) at ssc-vm-1624.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x54 (ios) at ssc-vm-1624.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x7d (ios) at ssc-vm-1625.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x8e (ios) at ssc-vm-1625.colo.seagate.com...
Stopped m0d@0x7200000000000001:0x1a (ios) at ssc-vm-1623.colo.seagate.com
Stopped m0d@0x7200000000000001:0x9 (ios) at ssc-vm-1623.colo.seagate.com
Stopped m0d@0x7200000000000001:0x54 (ios) at ssc-vm-1624.colo.seagate.com
Stopped m0d@0x7200000000000001:0x43 (ios) at ssc-vm-1624.colo.seagate.com
Stopped m0d@0x7200000000000001:0x7d (ios) at ssc-vm-1625.colo.seagate.com
Stopped m0d@0x7200000000000001:0x8e (ios) at ssc-vm-1625.colo.seagate.com
Stopping m0d@0x7200000000000001:0x2b (confd) at ssc-vm-1623.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x65 (confd) at ssc-vm-1624.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x9f (confd) at ssc-vm-1625.colo.seagate.com...
Stopped m0d@0x7200000000000001:0x65 (confd) at ssc-vm-1624.colo.seagate.com
Stopped m0d@0x7200000000000001:0x9f (confd) at ssc-vm-1625.colo.seagate.com
```

**Review by suvratjoshi — APPROVED** [review]

**Review by knekrasov — COMMENTED** [review]
@mssawant What will happen if the user invokes `hctl bootstrap` but Consul is not started (or if it is even not installed)?
Will it provide a human-readable error?

**Comment by vaibhavparatwar** [comment]
@mssawant  will this issue get addressed with this change?  https://jts.seagate.com/browse/EOS-20861 ? here its saying hare-consul-agent is in disabled state.

**Comment by mssawant** [comment]
@vaibhavparatwar, this PR is related to removing of hare-consul-agent,  https://jts.seagate.com/browse/EOS-20861 is different.
This PR does not address  https://jts.seagate.com/browse/EOS-20861.

**Comment by mssawant** [comment]
@knekrasov, 
>@mssawant What will happen if the user invokes hctl bootstrap but Consul is not started (or if it is even not installed)?
Will it provide a human-readable error?

Yeah, good point. We have a systemd dependency of consul.service in hare-hax.service but having a human readable message will help. Let me add that check. Thank you.

File: utils/hare-bootstrap [hunk 1]
```diff
@@ -293,62 +293,53 @@ EOF
...
-        ssh_error_info "$node" "sudo systemctl stop hare-consul-agent" &
-        pids+=($!)
-    done < <(get_client_nodes)
-    wait4 ${pids[@]-}
-    echo ' OK'
+say 'Checking Consul server on this node...'
+if ! sudo systemctl list-unit-files consul.service | \
+                          grep consul > /dev/null || \
+   ! sudo systemctl --quiet is-active consul.service; then
+    die "Consul not installed or running"
```
Comment (knekrasov) [inline utils/hare-bootstrap:300]: [minor]
```suggestion
    die "Consul is either not installed or not running"
```

**Review by knekrasov — APPROVED** [review]

File: utils/hare-bootstrap (hunk 1 above)
Comment (mssawant) [reply utils/hare-bootstrap]: Okay.

**Pull request merged**